include/boost/histogram/display.hpp

```diff
@@ -0,0 +1,212 @@
+                 [](int i) { return std::to_string(i); });
+  return string_values;
+}
+
+std::string draw_line(const unsigned int num, const char c = '*', bool complete = true) {
+  std::stringstream line;
+  unsigned int i = 0;
+  for (; i < num; ++i) line << c;
+
+  if (complete == true) {
+    for (; i < histogram_width; ++i) line << ' ';
+  }
+  return line.str();
+}
+
+std::string get_single_histogram_line(const std::vector<int>& values,
+                                      const unsigned int index) {
+  std::stringstream line;
+  line << "|" << draw_line(values.at(index)) << '|';
+  return line.str();
+}
+
+std::string get_external_line(const unsigned int labels_width) {
+  std::stringstream external_line;
+
+  external_line << draw_line(labels_width, ' ', false) << " +"
+                << draw_line(histogram_width, '-') << '+';
+  return external_line.str();
+}
+
+std::string get_top_line(const unsigned int labels_width,
+                         const std::vector<int>& values) {
+  std::stringstream top_line;
+
+  const unsigned int min = 0;
+  auto max_value = std::max_element(values.begin(), values.end());
+  const float max = *max_value / max_bin_coefficient;
+
+  std::string min_max = str(boost::format("%-i %.1f") % min %
+                            boost::io::group(std::setw(histogram_width), max));
+
+  top_line << draw_line(labels_width, ' ', false) << " " << min_max;
+  return top_line.str();
+}
+
+visualization_data precalculate_visual_data(extract& h_data){
+  const unsigned int additional_offset = 8; // 8 white characters  
+  const auto scale_factors = calculate_scale_factors(h_data.values_);
+  const auto str_values = convert_to_str_vec(h_data.values_);
+  const auto lower_width = get_max_width(h_data.lower_bounds_); 
+  const auto upper_width = get_max_width(h_data.upper_bounds_);
+  const auto str_values_width = get_max_width(str_values);
+  const auto hist_shift = lower_width + upper_width + str_values_width + additional_offset;
+
+  visualization_data v_data(str_values, scale_factors, lower_width, upper_width,
+                            str_values_width, hist_shift);
+  return v_data;
+}
+
+std::string draw_histogram(const extract& h_data, const visualization_data& v_data) {
+  std::stringstream visualisation;
+  
+  visualisation << "\n" << get_external_line(v_data.external_line_shift_) << "\n";
+
+  for (unsigned int i = 0; i < h_data.size(); i++)
+    visualisation << "  " 
+                  << get_single_label(h_data, i, v_data.lower_bounds_width_, v_data.upper_bounds_width_) << "  "
+                  << get_single_str_value(v_data.str_values_, i, v_data.str_values_width_) << " "
+                  << get_single_histogram_line(v_data.scale_factors_, i) << "\n";
+                   
+  visualisation << get_external_line(v_data.external_line_shift_) << "\n\n";
+  return visualisation.str();
+}
+} // namespace
+
+template <class histogram>
+void display(const histogram& h, std::ostream& output=std::cout) {
```

Comment:
This is not good design. You provide the same functionality as the ostream operator `template <class Histogram> std::ostream& operator<<(std::ostream& os, const Histogram& h)` but you don't use it. The idiomatic way to print things in C++ is the ostream operator and your implementation will replace the crude one that is currently implemented. :)

The only complication here is, that the ostream operator does not allow us to add some options. Therefore we use a little wrapper that allows us to pass an option together with the histogram.

```
namespace boost {
namespace histogram {
namespace detail {
// the main private display function
void display_histogram(std::ostream& os, const Histogram& h, unsigned terminal_width) {
   // ...
}
} // ns detail

template <class Histogram>
struct display_t {
   const Histogram& histogram;
   unsigned terminal_width;
};

template <class Histogram>
auto display(const Histogram& h, terminal_width=0) {
   return display_t<Histogram>{h, terminal_width == 0 ? 60 : terminal_width};
}

template <class Histogram>
std::ostream& operator<<(std::ostream& os, const display_t<Histogram>& dh) {
   // call your main display function here
   detail::display_histogram(os, dh.histogram, dh.terminal_width);
   return os;
}

template <class Histogram>
std::ostream& operator<<(std::ostream& os, const Histogram& h) {
   os << display(h);
   return os;
}

} // ns histogram
} // ns boost
```

So the user can either do `std::cout << hist << std::endl` or `std::cout << display(hist, 40) << std::endl`.

Reply:
@HDembinski You are right, thank you.
I would like to clarify requirements related to ostream operator.
At this time, basic ostream operator is defined in _ostream.hpp_, in namespace _boost::histogram_.

You mentioned previously that display.hpp should be "optional". 
Does it mean that after including display.hpp
`#include <boost/histogram/display.hpp>` _in include/boost/histogram.hpp_
 default ostream should be changed to "ascii ostream" which I am adding in _display.hpp_?
It would mean that existing ostream operator should be overwritten?

Defining two ostream operators in namespace _boost::histogram_ results results 
"ambiguous overload error". 

It seems to me that BOOST_TEST_NE requires ostream operator. 
After changing default ostream to "ascii version", I would like to force BOOST_TEST_NE to use below form for comparison: 
`histogram(regular(5, -0.5, 2, options=underflow | overflow))`.

What do you think about this approach?

Reply:
@HDembinski 
I have a problem with ostream operator. 
BOOST_TEST_EQ which I mentioned above is part of lightwest_test.hpp, so my previous idea is probably wrong. 
How can I solve properly ambiguouity? 
should I "encapsulate" ostream operator 
from histogram/include/boost/histogram/ostream.hpp
in new namespace?
Could you give me suggestion? 
I am also not sure why travis check didn't start :(